### ISSUE TITLE
Possibility of a "The handle is invalid" when reading files

### ISSUE THREAD

**Issue by alankila** [AgNO3/jcifs-ng#5]
This issue probably needs a little bit more visibility, which is why I'm opening this ticket. I have a production environment where I have this construct:

```
new ZipInputStream(new BufferedInputStream(file.getInputStream(), 1000000))
```

reading files from a share on a Windows server. These files are zip archives containing PDF files, from which I am extracting the TIFF streams and converting them into web formats (png/jpeg) for viewing using a browser UI. The conversion process can run for an hour depending on the size of the file -- up to  1 GB large have been spotted in the wild. However, when reading some zips, but not others, and seemingly without any good reason, I get the following stack trace after a few minutes since starting the read:

```
jcifs.smb.SmbException: The handle is invalid.
at jcifs.smb.SmbTransport.checkStatus(SmbTransport.java:762)
at jcifs.smb.SmbTransport.send(SmbTransport.java:896)
at jcifs.smb.SmbSession.send(SmbSession.java:194)
at jcifs.smb.SmbTree.send(SmbTree.java:156)
at jcifs.smb.SmbFile.send0(SmbFile.java:1012)
at jcifs.smb.SmbFile.send(SmbFile.java:974)
at jcifs.smb.SmbFile.send(SmbFile.java:966)
at jcifs.smb.SmbFileInputStream.readDirect(SmbFileInputStream.java:241)
at jcifs.smb.SmbFileInputStream.read(SmbFileInputStream.java:182)
```

The rest of the stack trace has been elided, as it is not from under jcifs.  This reading process can be restarted, but it will trigger again for the same file in a few minutes after making progress worth a few PDFs within that zip each time.

This issue has been reported with the original jcifs too, so I believe it is an issue with the actual codebase. I've not been able to reproduce this issue except on that one system, and being a production environment where customer actually loses money due to this issue, I had to solve it quickly. Because it is a Windows server using a Windows share, I am now using the native file share mapping using a drive letter, where IO appears to work correctly.

**Comment by mbechler** [comment]
Can you provide debug logs and/or packet traces for the situation you describe? The stacktrace alone unfortunately is not very helpful.

My best guess what could happen is:
- connection/session timeout occurs with your usage pattern ( with the large buffer size it might not perform a read for more than 35s ) 
- we fail to properly reopen the file after that and perform a read on the invalid handle, there is clearly a race condition in that code.

If that is the case, setting jcifs.smb.client.soTimeout=0 should help. Could you try that?

Unfortunately the idle timeout handling is a really big mess and needs a redesign. I don't think that the specific issue can be easily fixed without that.

**Comment by alankila** [comment]
Pardon me, what is this 35 second timeout, where is it? Can I increase it?

**Comment by mbechler** [comment]
Set system property jcifs.smb.client.soTimeout in milliseconds (0 should disable) 
(or Configuration.getSoTimeout() in the new config scheme)

**Comment by alankila** [comment]
I read the code in SmbSession and it comes up with this concept called expiration based on the soTimeout property. Now, I fully support the idea of requiring server to respond within a particular timeout, e.g. if I send a command, then the server should respond within 30 seconds.

But why on Earth does the SmbTransport destroy an idle connection after the soTimeout has elapsed, even in absence of any network commands? Intuitively, something called "socket timeout" is supposed to control for how long we are willing to wait server responses during actual network transactions. Near as I understand it, the sequence of commands is like this:

1. Issue a read.
2. Wait 40 seconds.
3. Issue a second read.
4. jcifs decides to destroy the session despite it knows that there's another read incoming.
5. Generate an error which nobody on Google seems to have a solution for.

Wouldn't the library be improved if the expiration property was removed altogether from SmbSession, and the code checking for expiration in SmbTransport would be removed too?  It seems to me that just setting sessionTimeout property to 0 would remove this misfeature, probably without impacting the desirable properties of having a socket timeout.

**Comment by mbechler** [comment]
Totally agree that it's broken. I already had found other concurrency issues like this one. 

It's not only the sessionTimeout but typically the SocketTimeout occuring in SmbTransport.peekKey(), that will cause the connection to close.

I don't think we can generally get rid of the session/connection idle timeouts or otherwise risk leaks (as long as users are not explicitly performing lifetime management using CIFSContext).

The idle disconnects shouldn't really be a problem per se (except for scenarios where the files is locked - but that would be another bug, it should be perfectly fine for that code to reestablish a session and reopen the file). It's mostly that they can happen at any point in time that causes a lot of trouble - e.g. between the open (still valid) and the read (then no longer valid). 
That could be fixed by having a semaphore that would prevent the transport to disconnect during critial sections. However that still doesn't fix the locking scenario. Let me think on that a bit further.

One quick fix might be an option to ignore SocketTimeout during peekKey if we were not actually waiting for something. That way we would retain socket timeout behaviour where it is desired while preventing the idle timeout disconnect. However we will lose the ability to disconnect idle connections.

Also the session timeout should probably be exposed as a system property so it can be easily disabled.

**Comment by mbechler** [comment]
Sorry for the rather long delay. 

Just pushed some changes that should allow to mitigate the issue (without totally disabling socket timeouts) if you set jcifs.smb.client.noIdleTimeout=true and jcifs.smb.client.sessionTimeout=0. Also managed to create a test case that somehwat reliably triggered an issue like the one you described.

As mentioned, a proper fix will require some bigger (possibly breaking) changes - it's on the list for an upcoming point release.

Would be great if you could check this does indeed fix (or rather mitigate) the issue you encountered?

**Comment by mbechler** [comment]
No feedback, but I belive this is properly fixed in 1.6/2.0.
